cobc/tree.c

```diff
@@ -2782,18 +2782,175 @@ cb_build_alphanumeric_literal (const void *data, const size_t size)
 }
 
 cb_tree
-cb_build_national_literal (const void *data, const size_t size)
+cb_build_gcos_literal(const void *data, const size_t size)
 {
 	cb_tree			l;
 
-	l = CB_TREE (build_literal (CB_CATEGORY_NATIONAL, data, size));
+	l = CB_TREE (build_literal (CB_CATEGORY_ALPHANUMERIC, data, size));
 
 	l->source_file = cb_source_file;
 	l->source_line = cb_source_line;
 
 	return l;
 }
 
+cb_tree
+cb_build_alphanumeric_literal (const void *data, const size_t size)
+{
+	cb_tree			l;
+
+#ifdef HAVE_ICONV_H
+	size_t outsize = size;
+	void * outdata = malloc(outsize);
+	memset(outdata, ' ', outsize);
+	
+	if(cb_iconv.alphanumeric == (iconv_t)-1) {
+		cobc_err_msg(_("iconv_open failed"));
+	} else{
+		size_t inbytesleft = size;
+		size_t outbytesleft = outsize;
+
+		char *inbuf = (char *)data;
+		char * outbuf = (char *)outdata;
+		
+		size_t convResult = iconv(cb_iconv.alphanumeric, &inbuf, &inbytesleft, &outbuf, &outbytesleft);
+		if(convResult == (size_t)-1) {
+			switch (errno) {
+			case E2BIG:
+				cobc_err_msg(_("iconv failed: Insufficient output buffer space"));
+				break;
+			case EILSEQ:
+				cobc_err_msg(_("iconv failed: Invalid multibyte sequence in the input"));
+				break;
+			case EINVAL:
+				cobc_err_msg(_("iconv failed: Incomplete multibyte sequence in the input"));
+				break;
+			default:
+				cobc_err_msg(_("iconv failed: Unknown error"));
+				break;
+			}
+
+		}
+
+		outsize -= outbytesleft;
+	}
+	l = CB_TREE (build_literal (CB_CATEGORY_ALPHANUMERIC, outdata, outsize));
+    free(outdata);
+#else
+    l = CB_TREE (build_literal (CB_CATEGORY_ALPHANUMERIC, data, size));
+#endif
+
+    l->source_file = cb_source_file;
+    l->source_line = cb_source_line;
+
+    return l;
+}
+
+cb_tree
+cb_build_national_literal (const void *data, const size_t size)
+{
+	cb_tree			l;
+	
+#ifdef HAVE_ICONV_H
+	size_t outsize = size*2;
+	void * outdata = malloc(outsize);
+	memset(outdata, ' ', outsize);
+	
+	if(cb_iconv.national == (iconv_t)-1) {
+		cobc_err_msg(_("iconv_open failed"));
+	} else{
+		size_t inbytesleft = size;
+		size_t outbytesleft = outsize;
+
+		char *inbuf = (char *)data;
+		char * outbuf = (char *)outdata;
+		
+		size_t convResult = iconv(cb_iconv.national, &inbuf, &inbytesleft, &outbuf, &outbytesleft);
+		if(convResult == (size_t)-1) {
+			switch (errno) {
+			case E2BIG:
+				cobc_err_msg(_("iconv failed: Insufficient output buffer space"));
+				break;
+			case EILSEQ:
+				cobc_err_msg(_("iconv failed: Invalid multibyte sequence in the input"));
+				break;
+			case EINVAL:
+				cobc_err_msg(_("iconv failed: Incomplete multibyte sequence in the input"));
+				break;
+			default:
+				cobc_err_msg(_("iconv failed: Unknown error"));
+				break;
+			}
+	    }
+
+        outsize -= outbytesleft;
+    }
+
+    l = CB_TREE (build_literal (CB_CATEGORY_NATIONAL, outdata, outsize));
+    free(outdata);
+#else
+    l = CB_TREE (build_literal (CB_CATEGORY_NATIONAL, data, size));
```

Comment:
"poor man's conversion" here?

Reply:
Why?

Reply:
the "poor man's" conversion that was originally in is still useful for the `!HAVE_ICONV` case ( = the `#else` if iconv is not available) - other than iconv it has to _assume_ that the source was encoded in an iso-8859 variant